test/e2e/run/report.go

```diff
@@ -0,0 +1,209 @@
+	}
+	if err != nil {
+		return E2EResult{}, fmt.Errorf("report file not found: %w", err)
+	}
+
+	e2eResult, err := e.parseJSONReport(reportPath)
+	if err != nil {
+		return e2eResult, fmt.Errorf("parsing ginkgo json report: %w", err)
+	}
+
+	return e2eResult, nil
+}
+
+func (e *E2EReport) parseJSONReport(reportPath string) (E2EResult, error) {
+	e2eResult := E2EResult{}
+	reportData, err := os.ReadFile(reportPath)
+	if err != nil {
+		return e2eResult, fmt.Errorf("reading report: %w", err)
+	}
+
+	var reports []ginkgoTypes.Report
+	err = json.Unmarshal(reportData, &reports)
+	if err != nil {
+		return e2eResult, fmt.Errorf("unmarshalling report: %w", err)
+	}
+
+	if len(reports) > 1 {
+		return e2eResult, fmt.Errorf("multiple reports found, expected 1")
+	}
+	report := reports[0]
+
+	e2eResult.TotalTests = report.PreRunStats.TotalSpecs
+
+	// Process each failed test
+	for _, spec := range report.SpecReports {
+		if spec.State == ginkgoTypes.SpecStateSkipped {
+			continue
+		}
+
+		if spec.LeafNodeType == ginkgoTypes.NodeTypeIt {
+			e2eResult.TestRan = e2eResult.TestRan + 1
+		}
+
+		// s3 artifacts path is set on the spec, try to find it to set at
+		// the root e2eresult level
+		artifactsPath := getReportEntry(spec, constants.TestArtifactsPath)
+		if artifactsPath != "" && e2eResult.ArtifactsBucketPath == "" {
+			// strip test name from path, which is the last part of the path
+			// ex: ?prefix=logs/<cluster-name>/<instance-name>/
+			path := strings.TrimSuffix(artifactsPath, "/")
+			e2eResult.ArtifactsBucketPath = path[:strings.LastIndex(path, "/")+1]
+		}
+
+		if spec.State == ginkgoTypes.SpecStatePassed {
+			continue
+		}
+
+		// This can be one of BeforeSuite, DeferCleanup and It
+		// if it failed, it will be included
+		// only Its will have log files created since ginkgo only captures stdout/stderr for It
+		leafType := spec.LeafNodeType
+		name := leafType.String()
+		if leafType == ginkgoTypes.NodeTypeIt {
+			name = strings.Join(spec.LeafNodeLabels, ", ")
+		}
+
+		// Get instance name from report entries
+		instanceName := getReportEntry(spec, constants.TestInstanceName)
+		failedTest := FailedTest{
+			InstanceName:   instanceName,
+			Name:           name,
+			State:          spec.State.String(),
+			FailureMessage: specFailureMessage(spec),
+		}
+
+		if artifactsPath != "" {
+			failedTest.GinkgoLog = filepath.Join(artifactsPath, testGinkgoOutputLog)
+			failedTest.SerialLog = filepath.Join(artifactsPath, constants.SerialOutputLogFile)
+		}
+		collectorLogsURL := getReportEntry(spec, constants.TestLogBundleFile)
+		if collectorLogsURL != "" {
+			failedTest.CollectorLogsBundle = collectorLogsURL
+		}
+
+		e2eResult.FailedTests = append(e2eResult.FailedTests, failedTest)
+
+		// Only process "It" test nodes for detailed logs
+		if spec.LeafNodeType != ginkgoTypes.NodeTypeIt {
+			continue
+		}
+
+		if failedTest.InstanceName == "" {
+			err = errors.Join(err, fmt.Errorf("no instance name found for test"))
```

Comment:
now that we don't have a defer, what do you think about keeping this in a slice and joining only on the return?